### DISCUSSION TITLE
Dam burning algorithm

### DISCUSSION THREAD

**Discussion by changliao1025** [changliao1025/pyflowline#54] [General]
The dam burning algorithm within pyflowline is added to include all the dam-associated flowlines in the workflow.
![dam_burning drawio](https://user-images.githubusercontent.com/20618384/199792971-3354ab67-f82e-4a77-a7f1-16c3aed189f0.png)

1. Obtain all the filtered flowline ID, A
2. Retrieve the upstream-downstream topology of all flowlines, B
3. Obtain the dam flowline list, C
4. Loop through C to get dam flowline ID, if it is in the A, skip, if not, find all downstream ID using B until reach A

**Comment by changliao1025** [comment]
In order to assign a flag to the direct downstream of a dam, the dam flowlines are classified into headwater and non-headwater list.

**Comment by changliao1025** [comment]
To make dam location and flowline consistent, we need to either:

- move the dam location to one of the vertices on flowline
- modify the flowline by adding a vertex at the dam location

**Reply by mgcooper** [reply]
Just an idea - but it might be worth discussing the idea of using control points to associate dams with flowlines rather than forcing them to align before we get too committed to one way of doing things. I am not clear on all of the reasons we want them to align, so perhaps I am missing something and we have no choice, but other models that include water management (e.g. ResSim) rely heavily on the concept of control points (basically a dictionary or set of key-value pairs) that associate each dam with one or more flowline vertices.

**Reply by changliao1025** [reply]
This lookup table is actually what is built-in in PyFlowline, dam and flowline are conceptually linked. However, it is possible they are located within the same cell in the final step.
